pixeltable/functions/video.py

```diff
@@ -221,66 +223,471 @@ def get_metadata(video: pxt.Video) -> dict:
+        if video_stream is None:
+            raise pxt.Error(f'concat_videos(): file {video} has no video stream')
+        resolutions.append((video_stream['width'], video_stream['height']))
+
+    # check for divergence
+    first_res = resolutions[0]
+    for i, res in enumerate(resolutions[1:], start=1):
+        if res != first_res:
+            raise pxt.Error(
+                f'concat_videos(): requires that all videos have the same resolution, but '
+                f'video 0 ({videos[0]}) has resolution {first_res[0]}x{first_res[1]} and '
+                f'video {i} ({videos[i]}) has resolution {res[0]}x{res[1]}.'
+            )
+
+    # ffmpeg -f concat needs an input file list
+    filelist_path = TempStore.create_path(extension='.txt')
+    with filelist_path.open('w', encoding='utf-8') as f:
+        for video in videos:
+            f.write(f"file '{video}'\n")
+
+    output_path = TempStore.create_path(extension='.mp4')
+
+    try:
+        # First attempt: fast copy without re-encoding (works for compatible formats)
+        cmd = ['ffmpeg', '-f', 'concat', '-safe', '0', '-i', str(filelist_path), '-c', 'copy', '-y', str(output_path)]
+        _logger.debug(f'concat_videos(): {" ".join(cmd)}')
+        try:
+            _ = subprocess.run(cmd, capture_output=True, text=True, check=True)
+            return str(output_path)
+        except subprocess.CalledProcessError:
+            # Expected for mixed formats - continue to fallback
+            pass
+
+        # we might some corrupted output
+        if output_path.exists():
+            output_path.unlink()
+
+        # general approach: re-encode with -f filter_complex
+        cmd = ['ffmpeg']
+        for video in videos:
+            cmd.extend(['-i', video])
+
+        all_have_audio = all(av_utils.has_audio_stream(str(video)) for video in videos)
+        video_inputs = ''.join([f'[{i}:v:0]' for i in range(len(videos))])
+        # concat video streams
+        filter_str = f'{video_inputs}concat=n={len(videos)}:v=1:a=0[outv]'
+        if all_have_audio:
+            # also concat audio streams
+            audio_inputs = ''.join([f'[{i}:a:0]' for i in range(len(videos))])
+            filter_str += f';{audio_inputs}concat=n={len(videos)}:v=0:a=1[outa]'
+        cmd.extend(['-filter_complex', filter_str, '-map', '[outv]'])
+        if all_have_audio:
+            cmd.extend(['-map', '[outa]'])
+
+        video_encoder = Env.get().default_video_encoder
+        if video_encoder is not None:
+            cmd.extend(['-c:v', video_encoder])
+        if all_have_audio:
+            cmd.extend(['-c:a', 'aac'])
+        cmd.extend(['-pix_fmt', 'yuv420p', str(output_path)])
+
+        _ = subprocess.run(cmd, capture_output=True, text=True, check=True)
+        return str(output_path)
+
+    except subprocess.CalledProcessError as e:
+        _handle_ffmpeg_error(e)
+    finally:
+        filelist_path.unlink()
+
+
+@pxt.udf(is_method=True)
+def overlay_text(
```

Comment:
I wonder if, rather than inventing our own bespoke parameter list, it wouldn't be better to just take a string as input that gets passed to `drawtext`? By introducing our own bespoke API, we're abstracting away the ffmpeg tool, which somewhat goes against our ethos ("don't abstract away the tools")

Then the signature would be:
```
def overlay_text(
    video: pxt.Video,
    text: str,
    *,
    drawtext: str
)
```
The docstring could give lots of clear examples on how to map typical patterns onto ffmpeg language.

Even if we keep all the bespoke parameters, I'd argue we should provide a way for users to specify their own `drawtext` instead; otherwise we're masking the full power of the underlying API.

Reply:
I have a hard time believing that this would be very popular. Moviepy didn't take that path, and it's a popular package. I would suggest we use this and have users tell us where it breaks down in their workflow.

Reply:
Ok, I'm fine with following the precedent set by moviepy.